File: apps/api/src/modules/proxy/rate-limiter.ts
import type { Redis } from "ioredis";
import { RateLimiterRedis } from "rate-limiter-flexible";
import { RateLimitError } from "@/lib/errors";

const limiters = new Map<string, RateLimiterRedis>();

const getLimiter = (
  redis: Redis,
  prefix: string,
  points: number,
  duration: number
): RateLimiterRedis => {
  const key = `${prefix}:${points}:${duration}`;
  let limiter = limiters.get(key);
  if (!limiter) {
    limiter = new RateLimiterRedis({
      duration,
      keyPrefix: prefix,
      points,
      storeClient: redis
    });
    limiters.set(key, limiter);
  }
  return limiter;
};

const SECONDS_PER_DAY = 86_400;

export const checkRateLimit = async (
  redis: Redis,
  keyId: string,
  rpm: number | null,
  rpd: number | null
): Promise<void> => {
  if (rpm === null && rpd === null) return;

  const checks = [
    ...(rpm === null
      ? []
      : [
          getLimiter(redis, "rl:rpm", rpm, 60)
            .consume(keyId)
            .then(
              () => {},
              () => {
                throw new RateLimitError(
                  "Rate limit exceeded (requests per minute)"
                );
              }
            )
        ]),
    ...(rpd === null
      ? []
      : [
          getLimiter(redis, "rl:rpd", rpd, SECONDS_PER_DAY)
            .consume(keyId)
            .then(
              () => {},
              () => {
                throw new RateLimitError(
                  "Rate limit exceeded (requests per day)"
                );
              }
            )
        ])
  ];

  await Promise.all(checks);
};

File: apps/api-py/src/proxy/rate_limiter.py
from redis.asyncio import Redis

from src.lib.errors import AppError


async def check_rate_limit(
    redis: Redis, key_id: str, rpm: int | None, rpd: int | None
) -> None:
    if rpm:
        rpm_key = f"raven:rl:rpm:{key_id}"
        current = await redis.incr(rpm_key)
        if current == 1:
            await redis.expire(rpm_key, 60)
        if current > rpm:
            raise AppError(
                f"Rate limit exceeded: {rpm} requests per minute",
                429,
                "RATE_LIMIT_EXCEEDED",
            )

    if rpd:
        rpd_key = f"raven:rl:rpd:{key_id}"
        current = await redis.incr(rpd_key)
        if current == 1:
            await redis.expire(rpd_key, 86400)
        if current > rpd:
            raise AppError(
                f"Rate limit exceeded: {rpd} requests per day",
                429,
                "RATE_LIMIT_EXCEEDED",
            )
